Add initial subset of `exprt` to internal `smt_termt` conversions

thomasspriggs opened the pull request:
Add initial subset of `exprt` to internal `smt_termt` conversions

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

thomasspriggs committed:
Add initial external interface for expr to smt conversion

This will likely need additional parameters for state and/or
configuration. These can be added in a following commit if and when they
are needed.

thomasspriggs committed:
Add place holder expr case switching in expr to smt conversion

Note that checks for `ID_object_size`, `ID_pointer_offset` and
`ID_pointer_object` are commented out due to missing classes for
representing these expressions.

thomasspriggs committed:
Add type_try_dynamic_cast support for `bool_typet`

thomasspriggs committed:
Add conversion of bool literal exprts to smt terms

thomasspriggs committed:
Add conversion of logical exprt operators to smt terms

thomasspriggs committed:
Add conversion of equivalence exprt operators to smt terms

thomasspriggs committed:
Add conversion of if_exprt operator to smt terms

thomasspriggs committed:
Add typet to smt sort conversion

thomasspriggs committed:
Add conversion of bit_vector literals to smt terms

The previous implementation duplicated the type to sort conversion from
`convert_type_to_smt_sort`, in checking the type of the expression. By
converting the type to a sort and then specialising based on the sort
rather than the type, this duplication is eliminated.
`convert_type_to_smt_sort` now defines the kind of encoding used for any
given expression.

thomasspriggs committed:
Add subclasses of binary_predicate_exprt for common relations

To facilite usage of these expressions in a type safe manner and without
directly refering to the id.

thomasspriggs committed:
Add conversion of relational operators to smt terms

TGWDB reviewed (commented):
Looks good so far (not going to approve as it is still draft).

thomasspriggs committed:
Remove dependency on C frontend

Front end specific features should have been lowered into non-language
specific operations before solving.

TGWDB commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -56,9 +56,15 @@ static smt_termt convert_expr_to_smt(const union_exprt &union_construction)
-    "Generation of SMT formula for constant literal expression: " +
-    constant_literal.pretty());
+    "Generation of SMT formula for constant literal expression of type: " +
+    constant_literal.type().pretty());
```
I thought the literal would include the type information? Would it be more helpful to know both type and literal value when debugging?

thomasspriggs replied:
The type is enough to determine the proximate cause of the invariant violation and is more specific to the condition this relates to in this case. This is because conversion of constant literal expressions are now (partially) implemented. This message is actually moved inside of `convert_type_to_smt_sort` instead in a subsequent refactor in this PR, where the full expression is unavailable. I would argue that if we are seriously engaged in debugging then we can add whatever additional printing is useful as part of that debugging process.
